service/rest_version.go

```diff
@@ -118,89 +119,142 @@ func copyVersion(srcVersion *model.Version, destVersion *restVersion) {
 			Author:   version.Author,
 			Revision: version.Revision,
 			Message:  version.Message,
 			Builds:   make(versionByBuild),
+			//Tasks:    make(versionStatusByTask),
 		}
 
 		result.Versions = append(result.Versions, versionInfo)
 	}
+	// Find all builds/tasks corresponding the set of version ids
+	if err = result.populateBuildsAndTasks(versionIds, versionIdx); err != nil {
+		msg := fmt.Sprintf("Error populating builds/tasks for recent versions of project '%v'", projectId)
+		grip.Errorf("%v: %+v", msg, err)
+		gimlet.WriteJSONInternalError(w, responseError{Message: msg})
+	}
 
-	for _, build := range builds {
-		buildInfo := versionBuildInfo{
-			Id:    build.Id,
-			Name:  build.DisplayName,
-			Tasks: make(versionByBuildByTask, len(build.Tasks)),
-		}
-
-		for _, task := range build.Tasks {
-			buildInfo.Tasks[task.DisplayName] = versionStatus{
-				Id:        task.Id,
-				Status:    task.Status,
-				TimeTaken: task.TimeTaken,
-			}
+	// create a page header
+	if nextPageStart != "" {
+		responder := gimlet.NewResponseBuilder()
+		err = responder.SetPages(&gimlet.ResponsePages{
+			Next: &gimlet.Page{
+				Relation:        "next",
+				LimitQueryParam: "limit",
+				KeyQueryParam:   "start",
+				BaseURL:         restapi.GetSettings().ApiUrl,
+				Key:             nextPageStart,
+				Limit:           l,
+			},
+		})
+		if err != nil {
+			msg := "error setting pages"
+			grip.Errorf("%v: %+v", msg, err)
+			gimlet.WriteJSONInternalError(w, responseError{Message: msg})
 		}
+		w.Header().Set("Link", responder.Pages().GetLinks(r.URL.String()))
+	}
+	gimlet.WriteJSON(w, result)
+}
 
-		versionInfo := result.Versions[versionIdx[build.Version]]
-		versionInfo.Builds[build.BuildVariant] = buildInfo
+func (r *recentVersionsContent) populateBuildsAndTasks(versionIds []string, versionIdx map[string]int) error {
+	builds, err := build.Find(
+		build.ByVersions(versionIds).
+			WithFields(build.BuildVariantKey, build.DisplayNameKey, build.TasksKey, build.VersionKey))
+	if err != nil {
+		return errors.Wrap(err, "Error finding recent versions")
+	}
+	tasks, err := task.Find(task.ByVersions(versionIds).
+		WithFields(task.IdKey, task.DisplayNameKey, task.StatusKey, task.TimeTakenKey, task.VersionKey, task.BuildVariantKey))
+	if err != nil {
+		return errors.Wrap(err, "Error finding recent tasks for recent versions")
 	}
 
-	gimlet.WriteJSON(w, result)
+	for _, b := range builds {
+		buildInfo := versionBuildInfo{
+			Id:    b.Id,
+			Name:  b.DisplayName,
+			Tasks: make(versionByBuildByTask),
+		}
+		versionInfo := r.Versions[versionIdx[b.Version]]
+		versionInfo.Builds[b.BuildVariant] = buildInfo
+	}
+	for _, t := range tasks {
+		taskInfo := versionStatus{
+			Id:        t.Id,
+			Status:    t.Status,
+			TimeTaken: t.TimeTaken,
+		}
+		// save task with the corresponding build for the corresponding version
+		versionInfo := r.Versions[versionIdx[t.Version]]
+		buildInfo := versionInfo.Builds[t.BuildVariant]
+		if buildInfo.Tasks == nil {
```

Comment:
This case is if that build variant wasn't seen in the build loop before, so tasks was never initialized. I think this shouldn't happen? But seems good to be cautious. Should we grip a warning? Because of how grip is used elsewhere in this file I wasn't sure if it would fit.